### ISSUE TITLE
machine_batch fails while machine alone resouce works fine

### ISSUE THREAD

**Issue by BGmot** [chef/chef-provisioning-aws#533]
I think I am missing something very basic here and any help would be appreciated.
The code below woks without any problem:
```
require 'chef/provisioning/aws_driver'
with_driver 'aws'
machine 'ey-test' do
  machine_options( {
    bootstrap_options: {
      associate_public_ip_address: false
    }
  } )
end
```

Following code fails (obviously I'd like to spin up more machines but even this simplistic test fails)
```
require 'chef/provisioning/aws_driver'
with_driver 'aws'
machine_batch do
  machine 'ey-test' do
    machine_options(
      { bootstrap_options: {
          associate_public_ip_address: false
        }
      }
    )
  end
end
```
chef-client's log:
```
  * machine_batch[default] action converge[2017-06-03T17:03:33-04:00] INFO: Processing machine_batch[default] action converge (kafka-cluster::test-aws line 3)
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/cheffish-5.0.1/lib/cheffish/merged_config.rb:6: warning: toplevel constant Mash referenced by Chef::Mash
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/cheffish-5.0.1/lib/cheffish/merged_config.rb:7: warning: toplevel constant Mash referenced by Chef::Mash
[2017-06-03T17:03:46-04:00] INFO: [Aws::EC2::Client 200 0.650918 0 retries] describe_instances(instance_ids:["i-06e40b1221bee104c"])

[2017-06-03T17:03:46-04:00] INFO: [Aws::EC2::Client 200 0.155659 0 retries] describe_instances(instance_ids:["i-06e40b1221bee104c"])

[2017-06-03T17:03:46-04:00] WARN: Machine ey-test (i-06e40b1221bee104c) is terminated.  Recreating ...
[2017-06-03T17:03:46-04:00] INFO: Processing aws_key_pair[chef_default] action create (basic_chef_client::block line 1385)
[2017-06-03T17:03:47-04:00] INFO: [AWS EC2 200 0.541622 0 retries] describe_key_pairs(:filters=>[{:name=>"key-name",:values=>["chef_default"]}])

[2017-06-03T17:03:47-04:00] INFO: [AWS EC2 200 0.252683 0 retries] describe_key_pairs(:key_names=>["chef_default"])

[2017-06-03T17:03:47-04:00] INFO: Processing directory[/home/eyurchen/.chef/keys] action create (basic_chef_client::block line 30)
[2017-06-03T17:03:47-04:00] INFO: Processing private_key[/home/eyurchen/.chef/keys/chef_default] action create (basic_chef_client::block line 104)

    - creating machine ey-test on aws
    -   max_count: 1
    -   min_count: 1
    -   instance_type: "t2.micro"
    -   image_id: "ami-2f06e04f"
    -   key_name: "chef_default"
    -   network_interfaces: [{"device_index"=>0, "associate_public_ip_address"=>false, "delete_on_termination"=>true}]
    ================================================================================
    Error executing action `converge` on resource 'machine_batch[default]'
    ================================================================================

    ArgumentError
    -------------
    parameter validator found 3 errors:
      - missing required parameter params[:image_id]
      - missing required parameter params[:min_count]
      - missing required parameter params[:max_count]

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/kafka-cluster/recipes/test-aws.rb

      3: machine_batch do
      4:   machine 'ey-test' do
      5:     machine_options(
      6:       { bootstrap_options: {
      7:           associate_public_ip_address: false
      8:         }
      9:       }
     10:     )
     11:   end
     12: end

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/kafka-cluster/recipes/test-aws.rb:3:in `from_file'

    machine_batch("default") do
      action [:converge]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      driver "aws"
      chef_server {:chef_server_url=>"http://localhost:7799", :options=>{:client_name=>"admin", :signing_key_filename=>"/home/eyurchen/.chef/chef-zero.pem", :api_version=>"0"}}
      declared_type :machine_batch
      cookbook_name "kafka-cluster"
      recipe_name "test-aws"
      machines ["ey-test"]
    end

    Platform:
    ---------
    x86_64-linux
```
Stacketrace:
```
$ less /var/chef/cache/chef-stacktrace.out 
Generated at 2017-06-03 17:03:47 -0400
ArgumentError: machine_batch[default] (kafka-cluster::test-aws line 3) had an error: ArgumentError: parameter validator found 3 errors:
  - missing required parameter params[:image_id]
  - missing required parameter params[:min_count]
  - missing required parameter params[:max_count]
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/param_validator.rb:28:in `validate!'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/param_validator.rb:13:in `validate!'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/plugins/param_validator.rb:20:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/plugins/logging.rb:39:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/seahorse/client/plugins/raise_response_errors.rb:14:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/plugins/idempotency_token.rb:18:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/aws-sdk-core/plugins/param_converter.rb:20:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/seahorse/client/plugins/response_target.rb:21:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/seahorse/client/request.rb:70:in `send_request'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-core-2.9.3/lib/seahorse/client/base.rb:207:in `block (2 levels) in define_operation_methods'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-resources-2.9.3/lib/aws-sdk-resources/request.rb:24:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-resources-2.9.3/lib/aws-sdk-resources/operations.rb:41:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-resources-2.9.3/lib/aws-sdk-resources/operations.rb:87:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/aws-sdk-resources-2.9.3/lib/aws-sdk-resources/operation_methods.rb:19:in `block in add_operation'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb:1491:in `block in create_instance_and_reference'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/retryable-2.0.4/lib/retryable.rb:68:in `retryable'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb:1484:in `create_instance_and_reference'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb:1445:in `block (2 levels) in create_servers'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36/lib/chef/chef_fs/parallelizer/parallel_enumerable.rb:263:in `process_input'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36/lib/chef/chef_fs/parallelizer/parallel_enumerable.rb:253:in `process_one'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36/lib/chef/chef_fs/parallelizer.rb:92:in `call'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-12.19.36/lib/chef/chef_fs/parallelizer.rb:92:in `worker_loop'
/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/logging-2.2.0/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

If I put following `fix` everything works well:
```
$ diff -ru /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb.fixed 
--- /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb        2017-06-03 17:01:51.190375287 -0400
+++ /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/chef-provisioning-aws-2.2.0/lib/chef/provisioning/aws_driver/driver.rb.fixed  2017-06-02 16:20:40.025285229 -0400
@@ -1442,6 +1442,8 @@
             machine_options = specs_and_options[machine_spec]
 
             clean_bootstrap_options = Marshal.load(Marshal.dump(bootstrap_options))
+            clean_bootstrap_options = deep_symbolize_keys(clean_bootstrap_options)
             instance = create_instance_and_reference(clean_bootstrap_options, action_handler, machine_spec, machine_options)
             converge_ec2_tags(instance, machine_options[:aws_tags], action_handler)
```
My versions
```
$ chef --version
Chef Development Kit Version: 1.3.43
chef-client version: 12.19.36
delivery version: master (dd319aa632c2f550c92a2172b9d1226478fea997)
berks version: 5.6.4
kitchen version: 1.16.0
$ gem list provisioning

*** LOCAL GEMS ***

chef-provisioning (2.2.1)
chef-provisioning-aws (2.2.0)
```

**Comment by robertomoutinho** [comment]
Having the same problem... machine_batch stopped working

**Comment by BenLiyanage** [comment]
I'm also having this issue... testing your PR right now

**Comment by BenLiyanage** [comment]
The change in this PR seems to atleast get me a little farther in the provisioning process, but I am still getting stuck with a subnet error: `Aws::EC2::Errors::InvalidSubnetIDNotFound`

**Comment by BGmot** [comment]
Hey @BenLiyanage! that does not seem related to chef-provisioning, it's rather a response from AWS API, double check all networking options you pass to your machine.

**Comment by BenLiyanage** [comment]
Yup, my settings were ok.  Things were getting (de)serialized incorrectly.

This PR fixes all my issues:
https://github.com/chef/chef-provisioning-aws/pull/532

**Comment by tyler-ball** [comment]
Should be fixed by #532 - LMK if it does not!
